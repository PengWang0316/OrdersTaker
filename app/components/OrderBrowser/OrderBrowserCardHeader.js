import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { CardHeader, Tooltip } from '@material-ui/core';
import { lightGreen, yellow, grey, red } from '@material-ui/core/colors';
import { CheckCircle, OfflinePin, WatchLater, Help } from '@material-ui/icons';

import { ORDER_STATUS_FINISHED, ORDER_STATUS_RECEIVED, ORDER_STATUS_PREPARING } from '../../config';

const avatarIcon = {
  width: 40,
  height: 40
};

const styles = {
  checkCircle: {
    ...avatarIcon,
    color: lightGreen[700]
  },
  offlinPin: {
    ...avatarIcon,
    color: yellow[800]
  },
  watchLater: {
    ...avatarIcon,
    color: grey[500]
  },
  unknow: {
    ...avatarIcon,
    color: red[500]
  }
};

export const OrderBrowserCardHeader = ({ order, classes, ...rest }) => {
  const orderDate = new Date(order.dateStamp);
  let avatarElement;
  switch (order.status) {
    case ORDER_STATUS_FINISHED:
      avatarElement = (
        <Tooltip title="Finished" placement="right-end">
          <CheckCircle className={classes.checkCircle} />
        </Tooltip>
      );
      break;
    case ORDER_STATUS_RECEIVED:
      avatarElement = (
        <Tooltip title="Received" placement="right-end">
          <OfflinePin className={classes.offlinPin} />
        </Tooltip>
      );
      break;
    case ORDER_STATUS_PREPARING:
      avatarElement = (
        <Tooltip title="Preparing" placement="right-end">
          <WatchLater className={classes.watchLater} />
        </Tooltip>
      );
      break;
    default:
      avatarElement = (
        <Tooltip title="Unknow Status" placement="right-end">
          <Help className={classes.unknow} />
        </Tooltip>
      );
      break;
  }
  return (
    <CardHeader
      avatar={avatarElement}
      title={orderDate.toDateString()}
      subheader={orderDate.toLocaleTimeString()}
      {...rest}
    />
  );
};
OrderBrowserCardHeader.propTypes = {
  order: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired
};
export default withStyles(styles)(OrderBrowserCardHeader);
