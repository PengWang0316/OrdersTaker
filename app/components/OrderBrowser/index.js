import OrderBrowser from './OrderBrowser';

export default OrderBrowser;
