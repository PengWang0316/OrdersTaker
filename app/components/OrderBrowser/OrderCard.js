import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { Card, CardMedia, CardContent, Typography, IconButton, Tooltip } from '@material-ui/core';
import { DeleteForever, Link } from '@material-ui/icons';
import { red } from '@material-ui/core/colors';

import OrderBrowserCardHeader from './OrderBrowserCardHeader';
import OrdersPageContext from '../../contexts/OrdersPageContext';

const styles = {
  card: {
    minWidth: 300,
    maxWidth: 500,
    margin: '10px 10px',
    '&:hover': {
      cursor: 'pointer'
    }
  },
  media: {
    height: 0,
    paddingTop: '56.25%', // 16:9
  },
  cardContent: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'space-between'
  },
  iconButton: {
    width: 30,
    height: 30
  },
  deleteButton: {
    color: red[400]
  },
  icon: {
    width: 20,
    height: 20
  }
};

export const OrderCard = ({ order, classes, menuItems }) => {
  const itemsKey = Object.keys(order.items);
  return (
    <OrdersPageContext.Consumer>
      {({ handleShowDetail, handleRmoveUnloginOrder, handleLinkOrder }) => (
        <Card className={classes.card}>
          <OrderBrowserCardHeader order={order} onClick={() => handleShowDetail(order)} />
          <CardMedia className={classes.media} image={menuItems[itemsKey[0]].photo} title={menuItems[itemsKey[0]].name} onClick={() => handleShowDetail(order)} />
          <CardContent className={classes.cardContent}>
            <div>
              <Typography className={classes.title} color="textSecondary">This order has {itemsKey.length} items</Typography>
              <Typography color="textSecondary">Total Price: ${order.totalPrice}</Typography>
            </div>
            {!order.userId && (
              <div>
                <Tooltip title="Remove this order from this device">
                  <IconButton onClick={() => handleRmoveUnloginOrder(order._id)} className={`${classes.iconButton} ${classes.deleteButton}`}><DeleteForever className={classes.icon} /></IconButton>
                </Tooltip>
                <Tooltip title="Link this order to your account">
                  <IconButton onClick={() => handleLinkOrder(order._id)} color="primary" className={classes.iconButton}><Link className={classes.icon} /></IconButton>
                </Tooltip>
              </div>
            )}
          </CardContent>
        </Card>
      )}
    </OrdersPageContext.Consumer>
  );
};
OrderCard.propTypes = {
  order: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired,
  menuItems: PropTypes.object.isRequired
};
export default withStyles(styles)(OrderCard);
