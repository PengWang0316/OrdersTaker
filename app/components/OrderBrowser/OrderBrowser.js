import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Typography } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import Pagination from '@kevinwang0316/react-materialui-pagination';
// import Pagination from '../Pagination';

import OrderCard from './OrderCard';
import { MAX_ORDER_AMOUNT } from '../../config';

const styles = {
  root: {
    width: '90%',
    margin: '40px auto'
  },
  title: {
    fontSize: 16,
    fontWeight: 'bold'
  },
  cardDiv: {
    display: 'flex',
    flexWrap: 'wrap'
  }
};

export const OrderBrowser = ({
  orders, classes, title, menuItems, offset, totalOrder, onPageNumberClick
}) => (
  <div className={classes.root}>
    <Typography color="primary" className={classes.title}>{title}</Typography>
    {Object.keys(menuItems).length !== 0 && (
      <div className={classes.cardDiv}>
        {orders.map(order => <OrderCard order={order} menuItems={menuItems} key={order._id} />)}
      </div>
    )}
    <Pagination
      offset={offset}
      limit={MAX_ORDER_AMOUNT}
      total={totalOrder}
      onClick={onPageNumberClick}
    />
  </div>
);
OrderBrowser.prototype = {
  orders: PropTypes.array.isRequired,
  classes: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired,
  menuItems: PropTypes.object.isRequired,
  onPageNumberClick: PropTypes.func.isRequired,
  offset: PropTypes.number.isRequired,
  totalOrder: PropTypes.number.isRequired
};
/* istanbul ignore next */
const mapStateToProps = state => ({
  menuItems: state.menuItems
});
export default connect(mapStateToProps, null)(withStyles(styles)(OrderBrowser));
