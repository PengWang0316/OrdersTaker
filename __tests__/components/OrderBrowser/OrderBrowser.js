import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { OrderBrowser } from '../../../app/components/OrderBrowser/OrderBrowser';

jest.mock('@material-ui/core/Card', () => 'Card');
jest.mock('@material-ui/core/CardMedia', () => 'CardMedia');
jest.mock('@material-ui/core/CardContent', () => 'CardContent');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/core/Tooltip', () => 'Tooltip');
jest.mock('@material-ui/icons/DeleteForever', () => 'DeleteForever');
jest.mock('@material-ui/icons/Link', () => 'Link');
jest.mock('../../../app/components/OrderBrowser/OrderBrowserCardHeader', () => 'OrderBrowserCardHeader');

describe('OrderBrowser', () => {
  const defaultProps = {
    classes: {
      root: 'root',
      title: 'title',
      cardDiv: 'cardDiv',
      card: 'card',
      media: 'media',
      cardContent: 'cardContent',
      iconButton: 'iconButton',
      deleteButton: 'deleteButton',
      icon: 'icon'
    },
    orders: [{
      _id: 'ordersId',
      items: {
        orderIdA: {},
        orderIdB: {}
      },
      userId: 'userId'
    }],
    menuItems: {
      orderIdA: { photo: 'orderA photo', name: 'orderA name' }
    },
    title: 'titel',
    onShowDetail: jest.fn(),
    onRemoveOrder: jest.fn(),
    onLinkOrder: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<OrderBrowser {...props} />);

  test('Snapshot with user id', () => expect(renderer.create(<OrderBrowser {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('Snapshot without user id', () => expect(renderer.create(<OrderBrowser {...{
    ...defaultProps,
    orders: [{
      _id: 'ordersId',
      items: {
        orderIdA: {},
        orderIdB: {}
      },
      userId: null
    }]
  }}
  />).toJSON()).toMatchSnapshot());

  test('Function called when click elements', () => {
    const newProps = {
      ...defaultProps,
      orders: [{
        _id: 'ordersId',
        items: {
          orderIdA: {},
          orderIdB: {}
        },
        userId: null
      }]
    };
    const component = getShallowComponent(newProps);
    component.find('OrderBrowserCardHeader').at(0).simulate('click');
    expect(defaultProps.onShowDetail).toHaveBeenCalledTimes(1);
    expect(defaultProps.onShowDetail).toHaveBeenLastCalledWith(newProps.orders[0]);
    component.find('CardMedia').at(0).simulate('click');
    expect(defaultProps.onShowDetail).toHaveBeenCalledTimes(2);
    expect(defaultProps.onShowDetail).toHaveBeenLastCalledWith(newProps.orders[0]);
    component.find('IconButton').at(0).simulate('click');
    expect(defaultProps.onRemoveOrder).toHaveBeenCalledTimes(1);
    expect(defaultProps.onRemoveOrder).toHaveBeenLastCalledWith(newProps.orders[0]._id);
    component.find('IconButton').at(1).simulate('click');
    expect(defaultProps.onLinkOrder).toHaveBeenCalledTimes(1);
    expect(defaultProps.onLinkOrder).toHaveBeenLastCalledWith(newProps.orders[0]._id);
  });
});
