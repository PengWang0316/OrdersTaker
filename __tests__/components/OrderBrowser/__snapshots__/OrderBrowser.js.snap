// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderBrowser Snapshot 1`] = `
<div
  className="root"
>
  <Typography
    className="title"
    color="primary"
  >
    titel
  </Typography>
  <div
    className="cardDiv"
  >
    <OrderCard
      menuItems={
        Object {
          "orderIdA": Object {
            "name": "orderA name",
            "photo": "orderA photo",
          },
        }
      }
      order={
        Object {
          "_id": "ordersId",
          "items": Object {
            "orderIdA": Object {},
            "orderIdB": Object {},
          },
          "userId": "userId",
        }
      }
    />
  </div>
  <Pagination
    limit={6}
    offset={0}
    onClick={[MockFunction]}
    total={100}
  />
</div>
`;
